Add feature to use version from annotated tag

shestee opened the pull request:
Previous plugin behaviour didn't allow missing `version.properties` file nor empty `version` property in the file. This commit adds a behaviour that allows such configuration and deducts version from tag that code is checked out on.
This functionality is useful for teams that prefer releasing "on demand" and also allows increasing version number as needed, not based on number of commits pushed as in default behaviour with `version` property set. To use this new simplified behaviour no other configuration option than missing `version` is needed.

As said implemented feature deducts version based on tag on which a code is checked out. If the tag is valid (has same prefix as configured or default _'v'_ and version number is valid) deducted version is same as version in tag. If other tags that are valid are found, `previousVersion` which is exposed by the plugin, is a version with highest number less than deducted version (eg. for tag `v1.1.5` plugin deducts version number `1.1.5` and assuming that tag `v1.1.4` exists, exposed `previousVersion` is `1.1.4`). If code is not checked out on a valid tag and `version` is not set as property the plugin deducts version `0.0.1`.

To implement this behaviour getter for `requestedVersion()` in `'VersionConfig` class was added, and it returns Optional, which is tested  in `NextVersionPicker` and `PreviousVersionFinder` classes for containing a value, and further logic in these classes is based upon this check for `requestedVersion`. Also `deductVersion()` method in `AutoVersion` checks if `requestedVersion` is found in configuration, and if not `previousVersion` is being found using newly deducted version which is passed as argument to `findPreviousVersion()`.
New behaviour is tested by two new unit tests in `NextVersionPickerTest`, two tests ("no file" and "missing version property") refactored in `VersionConfigTest` and was also tested manually by building test project with various configuration options.

This PR resolves #1

shestee committed:
Add feature to use version from annotated tag

Previous plugin behaviour didn't allow missing 'version.properties' file
nor empty 'version' property in the file. This commit adds a behaviour that
allows such configuration and deducts version from tag that code is checked out
on.
This functionality is useful for teams that prefer releasing 'on demand'
and also allows increasing version number as needed, not based on number of
commits pushed as in default behaviour with 'version' property set. To use this
new simplified behaviour no other configuration option than missing 'version'
is needed.

mockitoguy commented on src/main/java/org/shipkit/auto/version/AutoVersion.java:
```diff
@@ -40,20 +40,39 @@ DeductedVersion deductVersion(String projectVersion) {
                     config, projectVersion);
+
+            if (!config.getRequestedVersion().isPresent()) {
+                previousVersion = previousVersionFinder.findPreviousVersion(versions, new VersionConfig(nextVersion, config.getTagPrefix()));
```
Can we cover this condition in the unit test?

mockitoguy commented on src/main/java/org/shipkit/auto/version/NextVersionPicker.java:
```diff
@@ -27,13 +29,29 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
+                result = matcher.group().substring(config.getTagPrefix().length());
+                explainVersion(log, result, "deducted version based on tag: '" + config.getTagPrefix() + result + "'");
+            } else {
+                result = "0.0.1";
```
Can we get this code branch covered in the unit test?

mockitoguy commented on src/main/java/org/shipkit/auto/version/VersionConfig.java:
```diff
@@ -52,19 +56,25 @@ String newPatchVersion() {
-            throw new ShipkitAutoVersionException(
-                    "Please create file 'version.properties' with a valid 'version' property, " +
-                            "for example 'version=1.0.*'", e);
+            System.out.println("[shipkit-auto-version] Ignoring file '" + versionFile.getName()
```
Let's follow up with a little clean up for this, but let's not block this PR for this edge case.

mockitoguy commented on src/main/java/org/shipkit/auto/version/VersionConfig.java:
```diff
@@ -52,19 +56,25 @@ String newPatchVersion() {
-                    "  Correct examples: 'version=1.0.*', 'version=2.10.100'");
+        String v;
+        if (p.containsKey("version") && p.get("version").toString().trim().equals("")) {
+            v = null;
```
Looks like this code branch is not covered by unit test

mockitoguy reviewed (changes requested):
Hey! Looks great, please consider my feedback! Can you link the issue ticket to the PR description?

shestee committed:
Add unit tests

Increased code coverage of new feature that allows using version from annotated tag.

shestee commented:
Thank you @mockitoguy for your review! I added some unit tests to increase code coverage as you requested.

mockitoguy commented:
Looks great, thank you!!!
